indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java

```diff
@@ -518,20 +684,52 @@ public SeekableStreamSupervisor(
     this.useExclusiveStartingSequence = useExclusiveStartingSequence;
     this.dataSource = spec.getDataSchema().getDataSource();
     this.ioConfig = spec.getIoConfig();
+    this.dynamicAllocationTasksProperties = ioConfig.getDynamicAllocationTasksProperties();
+    log.info("Get dynamicAllocationTasksProperties from IOConfig : " + dynamicAllocationTasksProperties);
+
+    if (dynamicAllocationTasksProperties != null && !dynamicAllocationTasksProperties.isEmpty() && Boolean.parseBoolean(String.valueOf(dynamicAllocationTasksProperties.getOrDefault("enableDynamicAllocationTasks", false)))) {
+      log.info("EnableDynamicAllocationTasks for datasource " + dataSource);
+      this.enableDynamicAllocationTasks = true;
+    } else {
+      log.info("Disable Dynamic Allocate Tasks");
+      this.enableDynamicAllocationTasks = false;
+    }
+    int taskCountMax = 0;
+    if (enableDynamicAllocationTasks) {
+      this.metricsCollectionIntervalMillis = Long.parseLong(String.valueOf(dynamicAllocationTasksProperties.getOrDefault("metricsCollectionIntervalMillis", 10000)));
+      this.metricsCollectionRangeMillis = Long.parseLong(String.valueOf(dynamicAllocationTasksProperties.getOrDefault("metricsCollectionRangeMillis", 6 * 10 * 1000)));
+      int slots = (int) (metricsCollectionRangeMillis / metricsCollectionIntervalMillis) + 1;
+      log.info(" The interval of metrics collection is " + metricsCollectionIntervalMillis + ", " + metricsCollectionRangeMillis + " timeRange will collect " + slots + " data points at most.");
+      this.queue = new CircularFifoQueue<>(slots);
+      taskCountMax = Integer.parseInt(String.valueOf(this.dynamicAllocationTasksProperties.getOrDefault("taskCountMax", 8)));
```

Comment:
We have deployed this auto scale feature on PRD environment for half a year. And it works stably and efficiently. The default values of all parameters are the best practice values continuously adjusted according to the online conditions. And the max number of ingest tasks in our cluster is 8. Just in case, maybe 4 is more reasonable, avoiding using up all resources. And users can set a larger value if 4 is not satisfied.